tf2onnx/shape_inference.py

```diff
@@ -9,18 +9,219 @@
 from __future__ import print_function
 from __future__ import unicode_literals
 import logging
-import numpy as np
-from onnx import onnx_pb
+import os
+from distutils.version import LooseVersion
+
+import tensorflow as tf
 from tf2onnx import utils
 
 # pylint: disable=logging-not-lazy,missing-docstring,consider-swap-variables
 
 
 logger = logging.getLogger(__name__)
 
+
+def tf_cpp_infer_shape(tf_graph):
+    """Invoke tensorflow cpp shape inference by re-importing graph_def."""
+    # invoke c api if tf version is below 1.8
+    if utils.get_tf_version() < LooseVersion("1.8"):
+        os.environ["TF_C_API_GRAPH_CONSTRUCTION"] = "1"
+
+    graph_def = tf_graph.as_graph_def(add_shapes=True)
+    with tf.Graph().as_default() as inferred_graph:
+        tf.import_graph_def(graph_def, name="")
+    return inferred_graph
+
+
+def infer_shape_for_graph(tf_graph):
+    """
+    Infer shape for Tensorflow ops.
+    Tensorflow explicitly sets shape for some ops in python code, such as Switch, Merge and TensorArrayGather.
+    These shapes may be lost after freezing TF graph to graph_def without add_shapes=True.
+    To bring these shapes back, we implement our own shape inference for these control flow ops based on one assumption:
+    **outputs of Merge op have the same shape (at least the same rank) of its inputs**.
+    With this assumption, our shape inference can handle:
+        1. in tf.cond, outputs of two branches have the same rank.
+        2. in tf.while_loop, loop variables don't change their rank.
+    """
+    no_shape_updated = True
+    while no_shape_updated:
+        no_shape_updated = False
+        for o in tf_graph.get_operations():
+            updated = infer_shape_for_op(o)
+            if updated:
+                no_shape_updated = True

```

Comment:
i prefer to invoke tf infer first for whole graph (should skip when tf version > 1.5?)
then if there is still some missing shape, print a log and run legacy shape infer.